Feeds: Batch Triggers (#5205)
scrapy/scrapy: docs/topics/feed-exports.rst

the last file was delivered.

.. _batches:

Batches
=======

.. versionadded:: VERSION

Batches enable multiple sequential files to be generated for a feed instead of getting
a single file. This is useful in getting item deliveries earlier before the crawl ends.

Batching is controlled by a batch handler, which is defined through the 
:ref:`batch feed option <feed-options>`.
For information about enabling and configuring the default batch handler, see
:class:`~scrapy.extensions.batches.BatchHandler`. You can alternatively 
:ref:`create your own batch handler <custom-batch-handler>`.

If you enable batching, you must also :ref:`set feed URIs appropriately with placeholders<batch-uri>`.

.. _default-batch-handler:

Default Batch Handler
----------------------

.. autoclass:: scrapy.extensions.batches.BatchHandler

.. _custom-batch-handler:

Custom Batch Handler
--------------------

To use a custom batch handler, write a custom batch handler class and assign 
that class or its import string to the :ref:`batch feed option <feed-options>`.

Each batch handler is a class that must implement the following methods:

> .. method:: __init__(self, feed_options)

Reviewer: I think you need to add `:noindex:` to every `method`, `class`, and other API entries that do not describe something that is part of Scrapy, but an interface for users to implement themselves. For example:

```suggestion
.. method:: __init__(self, feed_options)
   :noindex:
```

Alternatively, you could wrap them as part of a class (e.g. `.. class:: MyBatchHandler`). I think that’s how we have done it in similar cases in the past. Although if `:noindex:` works, maybe that would be best. Up to you.

    Initialize the batch handler.

    :param feed_options: feed-specific :ref:`options <feed-options>`

    :type feed_options: dict

>     .. note:: Feed options support arbitrary parameters passing which, can be
>               used by custom batch handlers to access parameters passed by users.

Reviewer: ```suggestion
    .. note:: Feed options support arbitrary parameters passing, which can be
              used by custom batch handlers to access parameters passed by users.
```

.. method:: item_added(self)

    Called every time after an item is exported to the feed.
    

>     Return `True` if a new batch file should be created now, or `False` otherwise.

Reviewer: My bad :sweat_smile: 

```suggestion
    Return ``True`` if a new batch file should be created now, or ``False`` otherwise.
```

    You may use this method to update batch state information.

.. method:: new_batch(self, file)

    Called after each batch file is created, including the first batch file.

    You may reset any internal state of your batch handler.

    :param file: file pointer of new batch storage

    :type file: :class:`BinaryIO`

An example for custom batch handler::

    class CustomBatchHandler:

        def __init__(self, feed_options):

>             if "batch_divisible_by" not in feed_options or feed_options["batch_divisible_by"] <= 0
>                 raise NotConfigured

Reviewer: ```suggestion
            if feed_options.get("batch_divisible_by", 0) <= 0:
                raise NotConfigured
```

been ended.

Batching is controlled by a batch handler, which is defined through the 
:ref:`batch feed option <feed-options>`.
For information about enabling and configuring the default batch handler, see
:class:`~scrapy.extensions.batches.BatchHandler`. You can alternatively 
:ref:`create your own batch handler <custom_batch_handler>`.

If you enable batching, you must also set :ref:`feed URIs appropriately<batch-uri>` with placeholders.

.. _default_batch_handler:

Default Batch Handler
----------------------

.. autoclass:: scrapy.extensions.batches.BatchHandler

.. _custom_batch_handler:

Custom Batch Handler
--------------------

Each batch handler is a class that must implement the following methods:

.. method:: __init__(self, feed_options)

    Initialize the batch handler.

    Parameters:

    :param feed_options: feed-specific :ref:`options <feed-options>`
    :type feed_options: :class:`dict`

    Required attributes:

    * `self.batch_id` (:class:`int`): id number of current batch, must be initialised to 0.
    * `self.enabled` (:class:`bool`): must be `True` if a non-empty limit is passed, otherwise `False`.

.. method:: item_added(self)

>     Update batch state info. Will be called everytime an item is exported to feed.

Author: Moving internal data updating code to `should_trigger` will make `item_added` method useless. I guess I should remove `item_added` if we are doing all the updating in `should_trigger`?

>     Update batch state info. Will be called everytime an item is exported to feed.

Reviewer: I’m slightly inclined to keep it because I think it has some potential for extension in the future. I’m thinking that in the future (or maybe now?) we could add an `item` parameter to it, so that a batch handler can trigger batch switches based on item content.

I’m also thinking that we might want to eventually have a counterpart `before_item_added` method, which would be called before an item is added to a batch and get not only the `item` as a parameter, but possibly also the `data` that will be written into the file, as both things could also be used by a batch handler to determine whether or not to switch to a new batch.

Instead of removing `item_added`, maybe we could remove `should_trigger` and have `item_adapter` work the way `should_trigger` does, so instead of:

```python
self.batches[slot.uri_template].item_added()
if self.batches[slot.uri_template].should_trigger():
    ...
```

we could have:

```python
create_new_batch = self.batches[slot.uri_template].item_added()
if create_new_batch:
    ...
```

>     Update batch state info. Will be called everytime an item is exported to feed.

Author: Oh nice, this could also be helpful in implementing hard limit for file size.

that file uploaded to the feed URI.

If you want item delivery to start earlier when using one of these storage
backends, use :ref:`batches<batches>` to split the output items in multiple
files, based on a specified limit. That way, as soon as a file reaches the specified
limit, that file is delivered to the feed URI, allowing item delivery to start
way before the end of the crawl.

.. _batches:

Batches
=======

.. versionadded:: VERSION

Batches enable multiple sequential files to be generated for a feed instead of getting
a single file. This is useful in getting item deliveries earlier before the crawl has
been ended.

> To enable batching, you can use :ref:`criterias<batch_criterias>` such as ``batch_item_count``
> , ``batch_file_size``, ``batch_duration`` or you can create :ref:`your own batch
> handlers<custom_batch_handler>` to handle batch deliveries.

Reviewer: This is written from the perspective of a user that is using the default batch handler. While this will be true for most users, the documentation should be accurate.

What about:

```suggestion
Batching is controlled by a batch handler, which is defined through the 
:ref:`batch feed option <feed-options>`.

For information about enabling and configuring the default batch handler, see
:class:`~scrapy.extensions.batches.BatchHandler`. You can alternatively 
:ref:`create your own batch handler <custom_batch_handler>`.
```

Then you could create a section called something like “Built-in Batch Handler”, and use `.. autoclass:: scrapy.extensions.batches.BatchHandler` there to provide documentation about the handler directly from its docstring. Then you can place in its docstring all the information about enabling and configuring the default batch handler, including what you currently have under the “Batch Criterias” section below (by the way, it’s “criteria” even in plural), without implying that every batch handler is supposed to be enabled and configured the same way.

When generating multiple output files, you must use at least one of the following
placeholders in the feed URI to indicate how the different output file names are

> generated:

Reviewer: You can probably replace this whole part about batch feed URIs with a paragraph that simply states that, if you enable batching, you must update your feed URI accordingly, and then link to a new section that contains all the information here.

This is to avoid the most part of the introduction about batches being something as specific as the handling of the feed URI. It also allows us to link to this information from other parts of the documentation and from the outside.

Default Batch Handler
----------------------

> .. autoclass:: scrapy.extensions.batches.DefaultBatchHandler

Reviewer: I think it may be better to leave the previous name. The new name could be quite confusing if we switch the default value to a different one in the future.

    .. versionadded:: 2.3.0

-   ``batch_file_size``: falls back to
    :setting:`FEED_EXPORT_BATCH_FILE_SIZE`.

    .. versionadded:: VERSION

-   ``batch_duration``: falls back to
    :setting:`FEED_EXPORT_BATCH_DURATION`.

    .. versionadded:: VERSION

-   ``batch``: a custom batch handler to handle batch deliveries.

    .. versionadded:: VERSION

> 

Reviewer: In the documentation about the batch settings, it should be clear that they only apply if using the default batch handler.

If assigned an integer number higher than ``0``, Scrapy generates multiple output files
storing up to the specified number of items in each output file.

.. setting:: FEED_EXPORT_BATCH_FILE_SIZE

FEED_EXPORT_BATCH_FILE_SIZE
----------------------------

.. versionadded:: VERSION

Default: ``0B``

If assigned a size larger than ``0B``, Scrapy generates multiple output files based on
the specified size as a soft limit.

The size must be a string in the format <``SIZE``><``STORAGE-UNIT``>, where ``STORAGE-UNIT``
must be byte unit based on powers of 2(KiB, MiB, GiB, TiB) or powers of 10(kB, MB, GB, TB).

.. setting:: FEED_EXPORT_BATCH_DURATION

FEED_EXPORT_BATCH_DURATION
----------------------------

.. versionadded:: VERSION

Default: ``0:0:0``

If assigned a duration longer than ``0:0:0``, Scrapy generates multiple output files based on
the specified duration as a soft limit.

> Duration must be a string in the format ``hours:months:seconds``.

Reviewer: The settings documentation should also clarify the scope, the default batch handler.

You could also remove the settings from here, mention them in the docstring of the default batch handler as equivalents to the corresponding feed option keys (indicating which one has a higher priority over the other), and move the `.. setting:: …` anchors to right above the paragraph or list item that mentions each setting.

        Specifically:

        -   ``batch_id``: ID of the file batch. See
            :ref:`Batches<batches>`.

>             If a batch handler is not enabled or no batch trigger
>             is passed, ``batch_id`` is always ``1``.

Reviewer: What about something more generic?

```suggestion
            If batching is not enabled, ``batch_id`` is always ``1``.
```
